### DISCUSSION TITLE
How-to: create a standalone eleventy binary

### DISCUSSION THREAD

**Discussion by aral** [11ty/eleventy#4111] [Enhancement Queue]
I just made a little time to start playing with eleventy and wanted to see how easy it would be to package into a single binary to ease distribution. Turns out – very easy (with one niggle, see below):

1. Clone eleventy repo
    ```sh
    git clone https://github.com/11ty/eleventy
    ```
2. Install Nexe:
    ```sh
     npm i -g nexe
    ```
3. Compile eleventy:
    ```sh
    nexe -i cmd.js -o eleventy --resources=src/Engines/*
    ```

4. Move eleventy somewhere on your system path (e.g., on Linux):
   ```sh
   sudo mv eleventy /usr/local/bin

That’s it. Now you can run eleventy as usual. The difference being, you don’t even need Node or npm installed to do so. (The resulting binary on my machine was 97MB uncompressed, 29MB compressed (.tar.gz). 

The one niggle I noticed is that live reload does not seem to be triggering for content markdown files (it does work on other files, for example, for njk layout files). Haven’t had time to look into why.

Not sure if you want to provide compiled binaries as it seems (from a cursory glance), npm is pretty much required for working with eleventy (e.g., plugins, etc.) so please feel free to close if not useful.

**Comment by DirtyF** [comment]
Interesting, did you notice any significant change on build time?

**Comment by aral** [comment]
@DirtyF No difference in render times and no discernable difference in startup times (maybe a tiny 1/4-1/2 second lag on the compiled version… but nothing that stands out).

**Comment by danfascia** [comment]
This is actually quite a useful discovery because sometimes I have the need to use 11ty as an embedded system in my medical environment so I can truly see a use case for this where node_modules are frustrating.

**Comment by aral** [comment]
@danfascia I went a little further and [spiked a quick integration](https://source.ind.ie/site.js/app/tree/eleventy) with [Site.js](https://sitejs.org) but I’m blocked at the moment [by this merge request](https://github.com/11ty/eleventy/pull/376) (see [this issue](https://github.com/11ty/eleventy/issues/368)) making it impossible to include Eleventy in any app that accepts a path to operate on (the merge request limits Eleventy to running only within the current working directory).

I was able to integrate [Hugo](https://gohugo.io) into Site.js for the next release (it was trivial as it is a single binary and can operate on any path) but this is going to block me from potentially integrating Eleventy also (which is sad as I like what I see so far).

**Comment by zachleat** [comment]
@aral What is the specific blocker? Work for #368 has not been merged into Eleventy. #376 did not include any restrictions to the site output target, afaik

**Comment by aral** [comment]
@zachleat Ah, thanks for the clarification, I was confused, sorry. Apparently there are bugs in Eleventy’s path handling in general which I’ve now been able to reproduce using the eleventy-base-blog project and eleventy installed via npm. Will file a separate issue for that.

**Comment by aral** [comment]
@zachleat Just opened issues #937 and #939 which can be trivially reproduced using the eleventy-base-blog project and eleventy installed via npm.

There’s also #938, which seems related, but will be very difficult to reproduce as it relies on certain files on my machine (but hints to a behaviour that shouldn’t be exhibited).

Hope these help. Would love to hear your thoughts when you get a chance.

**Comment by Ryuno-Ki** [comment]
Thanks for the research, @aral.
I was wondering, whether somebody asked for packaging for GNU/Linux distributions?
For example, Debian or Arch Linux.
Perhaps they could use your findings in the process?

**Comment by zachleat** [comment]
Cross linking this: https://nodejs.org/api/single-executable-applications.html

**Comment by zachleat** [comment]
Let’s regroup at #3969 as the home base for this discussion!
